text/2018-09-17-schedule-limit.md

```diff
@@ -0,0 +1,176 @@
+an operator, it needs to guarantee that if we add all the schedule cost in the
+_runningOperators_, the total cost of any store will not exceed the configured
+value.
+
+If the first operator in the waiting queue makes a store overload, it will wait
+until the conflicting operator finishes or times out before it can be moved to
+_runningOperators_.
+
+In order to improve efficiency of executing, when a non-first operator does not
+conflict with any operator in front of it, it can be moved to
+_runningOperators_ too.
+
+### How to ensure fair competition between different schedulers
+
+This is actually quite straightforward. We only need to limit the number of
+generated operators by each scheduler in the _waitingOperators_ queue. For
+example, up to 3 operators from a same scheduler.
+
+Only when an old operator is moved into the _runningOperators_, the
+corresponding scheduler will have the opportunity to insert more operators into
+_waitingOperators_. This strategy not only ensures fairness between different
+schedulers, but also encourages schedulers to generate non-conflicting
+operators, thus reducing the impact on cluster performance.
+
+As before, if different schedulers generate operators of a same Region, the
+later one is rejected directly. Dropping the later operator when it is about
+to promote is also an option.
+
+### Configuration and customization
+
+There is only 1 important configuration for basic usage:
+
+* _MaxScheduleCost_
+
+    The maximum summary of cost of running operators for each store.
+
+We can also allow to customize following configurations to better fit special
+needs. (Tests are needed to decide if each of them is worth being made
+configurable):
+
+* _StoreMaxScheduleCost_
+
+    Similar to _MaxScheduleCost_, but for a specific store. If could be useful
+    when performance of `tikv-server`s are different.
+
+* _OperatorStepCost_
+
+    We will assign cost values based on experience. They may need adjustment
+    in production.
+
+* _MaxWaitingOperator_
+
+    The maximum count of waiting operators of each scheduler.
+
+* _SchedulerMaxWaitingOperator_
+
+    Similar to _MaxWaitingOperator_, but for a specific scheduler. We can use
+    it to control priority of different schedulers.
+
+## Drawbacks
+
+The migration from the old way to the new scheme has a certain cost, involving
+the update of the deployment script and compatibility with the old cluster
+configuration in the future.
+
+## Alternatives
+
+It has been considered that the schedulers use a cooperative rather than a
+competitive scheme to generate operators. However, this approach is more
+dependent on the better quality of implementation of all the schedulers, and
+the schedulers need to perceive each other, which will increase the coupling
+degree of the system.
+
+## Unresolved questions
+
+The arbitrarily determined step cost may not correctly reflect the scheduling
+overhead, we need to test to verify that it is appropriate.
```

Comment:
```suggestion
overhead, so we need to test to verify that it is appropriate.
```

Reply:
Note: Do not combine two simple sentences using a comma (,); instead, combine two simple sentences using a semicolon (;) or a conjunction (and, or, so, because and so on).